rllib/utils/metrics/stats/base.py

```diff
@@ -0,0 +1,291 @@
+import threading
+import time
+from abc import ABCMeta, abstractmethod
+from collections import deque
+from typing import Any, Dict, List, Optional, Union
+
+from ray.rllib.utils.annotations import (
+    OverrideToImplementCustomLogic_CallToSuperRecommended,
+)
+from ray.util.annotations import DeveloperAPI
+
+
+@DeveloperAPI
+class StatsBase(metaclass=ABCMeta):
+    """A base class for Stats.
+
+    Stats are meant to be used to log values to and then aggregate them in a tree.
+
+    Therefore, we log to stats in two different ways:
+    - On a leaf component, we log values directly by pushing.
+    - On a non-leaf component, we only aggregate incoming values.
+
+    Additionally, we pay special respect to Stats that live at the root of the tree.
+    These may have a different behaviour (example: a lifetime sum).
+
+    Note the tight coupling between StatsBase and MetricsLogger.
+    """
+
+    # In order to restore from a checkpoint, we need to know the class of the Stats object.
+    # This is set in the subclass.
+    stats_cls_identifier: str = None
+
+    def __init__(
+        self,
+        is_root: bool = False,
+        is_leaf: bool = True,
+    ):
+        """Initializes a StatsBase object.
+
+        Args:
+            is_root: If True, the Stats object is a root stats object.
+            is_leaf: If True, the Stats object is a leaf stats object.
+        """
+        self.is_root = is_root
+        self.is_leaf = is_leaf
+        # Used to keep track of start times when using the `with` context manager.
+        # This helps us measure times with threads in parallel.
+        self._start_times = {}
+        # For non-leaf stats (root or intermediate), track the latest merged values
+        # This is overwritten on each merge operation
+        if not self.is_leaf:
+            self.latest_merged: Union[List[Any], Any] = None
+
+        assert (
+            self.stats_cls_identifier is not None
+        ), "stats_cls_identifier must be set in the subclass"
+
+    @property
+    def has_throughputs(self) -> bool:
+        """Returns True if the Stats object has throughput tracking enabled.
+
+        Some Stats classes may have throughput tracking enabled, such as SumStats.
+        """
+        return False
+
+    def initialize_throughput_reference_time(self, time: float) -> None:
```

Comment:
Should this have something like `OverrideToImplementCustomLogic_CallToSuperRecommended` like `OverrideToImplementCustomLogic`?